kolena/workflow/_datatypes.py

```diff
@@ -44,6 +45,14 @@
 T = TypeVar("T", bound="DataObject")
 
 
+def _double_under(input: str) -> bool:
+    return input.startswith("__") and input.endswith("__")
+
+
+def allow_extra(cls) -> bool:
+    return "__pydantic_model__" in vars(cls) and cls.__pydantic_model__.Config.extra == Extra.allow
```

Comment:
Could use an extra pair of eye to see if there's any issue with this logic.

Reply:
just want to double check that if this is expected

```py
def test__data_object__extras_allow() -> None:
    class Config:
        extra = Extra.allow

    @dataclass(frozen=True, config=Config)
    class DataclassesTester(DataObject):
        b: float
        a: str
        z: bool

    @dataclass(frozen=True) 
    class Base(DataObject):
        ...

    @dataclass(frozen=True)
    class CustomDataInternal(Base):
        foo: str

    @dataclass(frozen=True)
    class CustomData(DataObject):
        foo: CustomDataInternal

    # extras should still be checked
    assert DataclassesTester(z=False, a="foobar", b=0.3, y=CustomData(foo=CustomDataInternal(foo="bar")))._to_dict() == {'a': 'foobar', 'b': 0.3, 'y': {'foo': {'foo': 'bar'}}, 'z': False}
``` 

```py
def test__data_object__extras_allow_invalid() -> None:
    class Config:
        extra = Extra.allow

    @dataclass(frozen=True, config=Config)
    class DataclassesTester(DataObject):
        b: float
        a: str
        z: bool

    @dataclass(frozen=True, config=Config) 
    class Base(dict):
        ...

    @dataclass(frozen=True)
    class CustomDataInternal(Base):
        foo: str

    @dataclass(frozen=True)
    class CustomData(DataObject):
        foo: CustomDataInternal

    # extras should still be checked
    assert DataclassesTester(z=False, a="foobar", b=0.3, y=CustomData(foo=CustomDataInternal(foo="bar")))._to_dict() == {'a': 'foobar', 'b': 0.3, 'y': {'foo': {}}, 'z': False}
```

Reply:
Given existing serialization logic, it is expected. `CustomDataInternal` is being detected as `dict`, which we then use `.items()` to serialize its content. Therefore `foo` is not considered.

It can be discussed of a better way to handle this case, however I'll leave that to the future.